File: CS315-Programming Languages/HW2/Codes/22101782_akin_yasemin.lua
--1
function outerFunction(x)
    return function(y) 
        return x + y
    end
end

local addFive = outerFunction(5)
print(addFive(3))  -- Output: 8

--2
function outerFunction2()
    local localVar = 10
    globalVar = 20
    local tableVar = { value = 30 }

    function innerFunction()
        return localVar, globalVar, tableVar.value
    end

    return innerFunction
end

local closure = outerFunction2()
print(closure()) -- Outputs: 10 20 30

--3
function outerFunction3()
    local outerVar = 10
    function innerFunction2()
        outerVar = outerVar + 5
        return outerVar
    end
    return innerFunction2
end

local myClosure = outerFunction3()
print(myClosure())  -- Output: 15
print(myClosure())  -- Output: 20

--4
function outerFunction4(outerVar2)
    function innerFunction3(innerVar2)
        print(outerVar2, innerVar2)
        return function(mostInnerVar)
            print(outerVar2, innerVar2, mostInnerVar)
        end
    end
    return innerFunction3
end

local nested = outerFunction4("Outer")("Inner")
nested("Most Inner")

File: CS315-Programming Languages/HW2/Codes/22101782_akin_yasemin.py
#1
def outer_function(x):
    def inner_function(y):
        return x * y
    return inner_function

multiplier_by_3 = outer_function(3)
print(multiplier_by_3(5))  # Output: 15

#2
def outer_function(x):
    y = 10
    global z
    z = 5

    def inner_function():
        nonlocal y
        y += 1
        print(f"x (local to outer): {x}, y (non-local): {y}, z (global): {z}")

    return inner_function

closure = outer_function(3)
closure()  # Output: x (local to outer): 3, y (non-local): 11, z (global): 5

#3
def outer():
    x = [10]

    def inner():
        nonlocal x
        x[0] += 1  # Modifies the list element, which is mutable
        return x[0]

    return inner

my_closure = outer()
print(my_closure())  # Output: 11
print(my_closure())  # Output: 12

#4
def outer_function(x):
    def middle_function(y):
        def inner_function(z):
            return x * y * z
        return inner_function
    return middle_function

nested_closure = outer_function(2)(3)
print(nested_closure(4))  # Output: 24